File: leetcode/combinations.cpp
#include "vector"
using namespace std;

class Solution
{
public:
    vector<vector<int>> array;
    vector<int> list;
    int tNum;
    vector<vector<int>> combine(int n, int k)
    {
        tNum = k;
        array.empty();
        list.empty();
        dfs(1, n);
        return array;
    }
    void dfs(int start, int end)
    {
        if (list.size() == tNum)
        {
            array.push_back(list);
            return;
        }
        if (list.size() + end - start + 1 < tNum)
            return;
        for (int i = start; i <= end; i++)
        {
            list.push_back(i);
            dfs(i + 1, end);
            list.pop_back();
        }
    }
};

File: leetcode/combinations.js
/**
 * 77. 组合
 * 给定两个整数 n 和 k，返回 1 ... n 中所有可能的 k 个数的组合。
 * @param {number} n
 * @param {number} k
 * @return {number[][]}
 */
let arrays = [];
let tNum = 0;
let subArray = [];
var combine = function (n, k) {
    arrays = [];
    subArray=[];
    tNum = k;
    handleArray(1, n);
    return arrays;
};
function handleArray(start, end) {
    if (subArray.length == tNum) {
        arrays.push(subArray.slice());
        return;
    }
    if (subArray.length + end - start + 1 < tNum) return;
    for (let i = start; i <= end; i++) {
        subArray.push(i);
        handleArray(i + 1, end);
        subArray.pop();
    }
}

console.log(combine(4, 2));